renderservice/render/daemon/RenderDaemon.py

```diff
@@ -0,0 +1,196 @@
+
+
+def authenticate_storage():
+    """Authenticate into Google Cloud storage.
+
+    Returns:
+        An authenticate storage client.
+    """
+    from google.auth import compute_engine
+    from google.cloud import storage
+
+    credentials = compute_engine.Credentials()
+    storage_client = storage.Client(credentials=credentials, project=PROJECT_NAME)
+
+    return storage_client
+
+
+def handle_timelimit_exceeded():
+    """Raise the timeout exception when SIGALRM signal is caught."""
+    raise TimeoutError("Timelimit exceeded.")
+
+
+class RenderDaemon(RunDaemon, ResourceGenerator):
+    """A daemon that processes tasks related to the rendering pipeline."""
+
+    def __init__(self, *args, **kwargs):
+        """Create a Render Daemon.
+
+        Assumes that any SIGALRM signals are sent by itself for
+        timeout exceptions.
+        """
+        super(RenderDaemon, self).__init__(*args, **kwargs)
+        self.handle(signal.SIGALRM, handle_timelimit_exceeded)
+        # Handle SIGUSR1 for closing up for pre-emption.
+
+    def run(self):
+        """Consumes jobs and produces rendered documents."""
+        queue = QueueHandler(project_name=PROJECT_NAME, taskqueue_name=QUEUE_NAME, discovery_url=DISCOVERY_URL)
+        logger.info("Daemon with pid {} running.".format(self.pid))
+        while True:
+            lease_secs = TASK_COUNT * TASK_SECONDS
+            tasks = queue.lease_tasks(tasks_to_fetch=TASK_COUNT, lease_secs=lease_secs, tag="task")
+            self.process_tasks(tasks, queue)
+            time.sleep(RENDER_SLEEP_TIME)
+
+    def process_tasks(self, tasks, queue):
+        """Run main loop for determining individual task logic.
+
+        Tasks will be run to recieve a result, saved if necessary
+        then deleted. Tasks that return a none result or are
+        interupted by an exception will not be deleted and have
+        their lease reduced for another daemon. Tasks that have
+        surpassed their retry limit will have a failure result
+        saved if necessary otherwise they will be deleted.
+
+        Args:
+            tasks: A list of json task objects.
+            queue: QueueHandler to update and delete tasks from.
+        """
+        for task_descriptor in tasks:
+            task_id = task_descriptor["id"]
+            retries = task_descriptor["retry_count"]
+            timeout_seconds = int(TASK_SECONDS + TASK_SECONDS * TASK_TIME_MULT * retries)
+
+            result = None
+            if retries < TASK_RETRY_LIMIT:
+                signal.alarm(timeout_seconds)
+                try:
+                    result = self.process_task(task_descriptor)
+                except Exception as e:
+                    logger.exception("Task {} raised exception with error: {}".format(task_descriptor["id"], e))
+                finally:
+                    signal.alarm(0)
+            else:
+                result = self.handle_retry_limit(task_descriptor)
+
+            # Save out documents
```

Comment:
"out documents" ?

Reply:
Fixed, was probably trying to capture the idea that they are not saved on the current machine.